ast/desugar/Desugar.cc

```diff
@@ -727,6 +748,26 @@ ExpressionPtr node2TreeImplBody(DesugarContext dctx, parser::Node *what) {
                     flags.isPrivateOk = true;
                 }
 
+                if (isCallToBlockGivenP(send, rec) && dctx.enclosingBlockParamName.exists()) {
+                    // Desugar:
+                    //     def foo(&my_block)
+                    //       x = block_given?
+                    //     end
+                    //
+                    // to:
+                    //     def foo(&my_block)
+                    //       x = (my_block ? ::Kernel.block_given?() : false)
+                    //     end
+                    //
+                    // Later stages of the pipeline have special handling for this,
+                    // based on the nilability of the block (if specified)
+
+                    auto sendExpr = MK::Send0(loc, move(rec), core::Names::blockGiven_p(), send->methodLoc);
```

Comment:
I'm on the fence about the receiver:

* I can either reuse the existing `rec` (fewer allocations)
* or I could hardcode a new `MK::Const` for `::Kernel` here (less variation in the desugar trees)